inmoose/limma/ebayes.py

```diff
@@ -0,0 +1,161 @@
+
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+# -----------------------------------------------------------------------------
+
+# This file is based on the file 'R/ebayes.R' of the Bioconductor limma package (version 3.55.1).
+
+
+import numpy as np
+import scipy
+
+
+def tmixture_matrix(tstat, stdev_unscaled, df, proportion, v0_lim=None):
+    """
+    Estimate scale factor in mixture of *t*-distributions
+
+    This function estimates the unscaled standard deviation of the true
+    (unobserved) log fold changes for differentially expressed genes. It is
+    used internally by :func:`eBayes` and is not intended to be called directly
+    by users.
+
+    The values in each column of :code:`tstat` are assumed to follow a mixture
+    of an ordinary *t*-distribution, with mixing proportion
+    :code:`1-proportion`, and :code:`(v0+v1)/v1` times a *t*-distribution, with
+    mixing proportion :code:`proportion`. Here :code:`v1` is
+    :code:`stdev_unscaled**2` and :code:`v0` is the value to be estimated.
+
+    Arguments
+    ---------
+    tstat : ndarray
+        matrix of *t*-statistics
+    stdev_unscaled : ndarray
+        matrix, conformal with :code:`tstat`, containing the unscaled standard
+        deviations of the coefficients used to compute the *t*-statistics
+    df : array_like
+        array giving the degrees of freedom associated with :code:`tstat`
+    proportion : float
+        assumed proportion of genes that are differentially expressed
+    v0_lim : pair of floats, optional
+        lower and upper limits for the estimated unscaled standard deviations
+
+    Returns
+    -------
+    ndarray
+        estimated :code:`v0` values, one for each column of :code:`tstat`
+    """
+    if tstat.shape != stdev_unscaled.shape:
+        raise ValueError("dimensions of tstat and stdev_unscaled do not match")
+    if (tstat.shape[0],) != df.shape:
+        raise ValueError("dimensions of tstat and df do not match")
+    if v0_lim is not None:
+        if len(v0_lim) != 2:
+            raise ValueError("v0_lim must be a pair")
+    ncoef = tstat.shape[1]
+    v0 = np.array(
+        [
+            tmixture_vector(tstat[:, j], stdev_unscaled[:, j], df, proportion, v0_lim)
+            for j in range(ncoef)
+        ]
+    )
+    return v0
+
+
+def tmixture_vector(tstat, stdev_unscaled, df, proportion, v0_lim=None):
```

Comment:
Same here can you add variable type ?

Reply:
types are documented in the docstring
I'm not sure that type annotation would add much info